### PR TITLE
feat: support multiple ADO organizations via unified project list

### LINKED ISSUE #186: Support multiple ADO organizations via unified project list
## Summary

Currently the ADO provider supports a single `workcenterAdo.organization` (string) plus a list of `workcenterAdo.projects` (string[]). This limits users to one ADO organization.

## Proposal

Consolidate the two settings into a single list, e.g. `workcenterAdo.projects` (string[]), where each entry is either:

- **`<org>`** — monitor an entire organization (no project filter)
- **`<org>/<project>`** — monitor a specific project within an organization

This allows users to work across multiple ADO organizations and freely mix org-wide and project-scoped entries.

### Example

```json
"workcenterAdo.projects": [
  "contoso",
  "fabrikam/ProjectA",
  "fabrikam/ProjectB"
]
```

## Migration

The existing `workcenterAdo.organization` setting would be deprecated. A migration path or backward-compatible fallback should be provided.

### PR DESCRIPTION
**Pull request by mthalman**
## Summary

Support multiple Azure DevOps organizations via a unified `workcenterAdo.projects` setting.

Each entry in `workcenterAdo.projects` can now be either:
- `<org>` — monitor an entire organization
- `<org>/<project>` — monitor a specific project within an organization

This allows users to monitor work items and PR reviews across multiple ADO organizations simultaneously.

## Changes

- **New `configParser.ts`** — Shared config parsing logic that converts the raw settings into per-org configurations.
- **Updated providers** — Both `AdoWorkItemProvider` and `AdoPrReviewProvider` now accept `OrgConfig[]` and iterate over all configured organizations. Per-org user ID caching added for the PR provider.
- **Updated `extension.ts`** — Uses the config parser to derive org configs from the settings. Warning messages distinguish between "nothing configured" and "all entries invalid".
- **Updated `package.json`** — Removed the deprecated `workcenterAdo.organization` setting. `workcenterAdo.projects` description updated to explain the new format.
- **`externalId` scoped to org** — Prevents collisions when the same project name exists in multiple organizations.
- **Terminal states cache scoped to org** — Prevents incorrect cache hits across orgs with different process templates.
- **Config parser tests** — Covers org/project parsing, org-only entries, edge cases (malformed entries, deduplication, whitespace).

## Breaking Change

The deprecated `workcenterAdo.organization` setting has been removed. Users must migrate to the new `workcenterAdo.projects` format using `org` or `org/project` entries.

Closes #186

### CODE REVIEW

File: packages/ado/src/adoPrReviewProvider.ts [hunk 1]
```diff
@@ -102,62 +101,61 @@ export class AdoPrReviewProvider extends BaseProvider {
...
-        logger.warn('Skipping invalid ADO project name', project);
+      const userId = await this.getUserId(accessToken, orgConfig.org, sessionAccountId);
+      if (!userId) {
+        const message = 'Failed to determine Azure DevOps user identity';
+        if (isUserTriggered) {
+          void vscode.window.showWarningMessage(`WorkCenter ADO: ${message}`);
+        }
+        logger.warn(message);
+        continue;
       }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPrReviewProvider.ts:119]: Fixed in 9e8fa02

File: packages/ado/src/configParser.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ * Parses the `workcenterAdo.projects` and (deprecated) `workcenterAdo.organization`
+ * settings into a list of per-organization configurations.
+ *
+ * New-format entries in `projects`:
+ *   - `<org>` — monitor an entire organization
+ *   - `<org>/<project>` — monitor a specific project
+ *
+ * Legacy backward compatibility: if no entry in `projects` contains `/` and
+ * `legacyOrganization` is non-empty, entries are treated as project names
+ * under that organization.
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/configParser.ts]: Fixed in dd0f351

File: packages/ado/src/adoWorkItemProvider.ts [hunk 3]
```diff
@@ -118,52 +117,52 @@ export class AdoWorkItemProvider extends BaseProvider {
...
+
+      results.forEach((result, index) => {
+        if (result.status === 'fulfilled') {
+          const { items, failed } = result.value;
+          allItems.push(...items);
+          if (failed) {
+            failures.push(projectList[index] || orgConfig.org);
+          }
+        } else {
+          const failureTarget = projectList[index] || orgConfig.org;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoWorkItemProvider.ts]: Fixed in 45ee7d5

File: packages/ado/src/adoPrReviewProvider.ts [hunk 4]
```diff
@@ -102,62 +101,58 @@ export class AdoPrReviewProvider extends BaseProvider {
...
-          failures.push(projectList[index] || this.org);
+      results.forEach((result, index) => {
+        if (result.status === 'fulfilled') {
+          const { items, failed } = result.value;
+          allItems.push(...items);
+          if (failed) {
+            failures.push(projectList[index] || orgConfig.org);
+          }
+        } else {
+          const target = projectList[index] || orgConfig.org;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPrReviewProvider.ts]: Fixed in 2985b76

File: packages/ado/src/adoWorkItemProvider.ts [hunk 5]
```diff
@@ -369,6 +371,7 @@ export class AdoWorkItemProvider extends BaseProvider {
    */
   private async filterActiveItems(
     token: string,
+    org: string,
     workItems: AdoWorkItem[],
   ): Promise<AdoWorkItem[]> {
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoWorkItemProvider.ts:379]: Fixed in d08f738

File: packages/ado/src/adoPrReviewProvider.ts [hunk 6]
```diff
@@ -102,62 +101,58 @@ export class AdoPrReviewProvider extends BaseProvider {
...
-    for (const project of this.projects) {
-      if (project === '' || isValidUrlSegment(project)) {
-        validProjects.push(project);
-      } else {
-        logger.warn('Skipping invalid ADO project name', project);
+      const userId = await this.getUserId(accessToken, orgConfig.org, sessionAccountId);
+      if (!userId) {
+        failures.push(`${orgConfig.org} (user identity)`);
+        logger.warn(`Failed to determine Azure DevOps user identity for org ${orgConfig.org}`);
+        continue;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPrReviewProvider.ts:118]: Fixed in f553553

File: packages/ado/src/adoWorkItemProvider.ts [hunk 7]
```diff
@@ -118,52 +117,54 @@ export class AdoWorkItemProvider extends BaseProvider {
   }
 
   private async fetchAndPublishWorkItems(accessToken: string, isUserTriggered: boolean): Promise<void> {
-    if (!isValidUrlSegment(this.org)) {
-      logger.warn('Skipping fetch: invalid ADO organization name', this.org);
-      this._onDidDiscoverItems.fire([]);
-      return;
-    }
+    const allItems: DiscoveredItem[] = [];
+    const failures: string[] = [];
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoWorkItemProvider.ts:122]: Fixed in 38347c9

File: packages/ado/src/configParser.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  // valid slash-separated entries (ignoring malformed ones like "/" or "/proj")
+  const hasNewFormatEntry = projects.some(p => {
+    const t = p.trim();
+    const idx = t.indexOf('/');
+    return idx > 0 && idx < t.length - 1;
+  });
+
+  // Legacy mode: no entry contains a valid '/' and the deprecated org setting is present
+  if (!hasNewFormatEntry && trimmedLegacyOrg) {
+    return [{ org: trimmedLegacyOrg, projects: projects.map(p => p.trim()).filter(Boolean) }];
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/configParser.ts]: Fixed in 5b15da7

File: packages/ado/src/adoPrReviewProvider.ts [hunk 9]
```diff
@@ -239,7 +249,7 @@ export class AdoPrReviewProvider extends BaseProvider {
     });
 
     if (!response.ok) {
-      logger.warn(`Failed to fetch PRs for project: ${project || this.org}`);
+      logger.warn(`Failed to fetch PRs for project: ${project || org}`);
       logger.error(`PR fetch failed for project "${project}": ${response.status}`);
       return { items: [], failed: true };
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPrReviewProvider.ts:255]: Fixed in 77b48c5

File: packages/ado/src/adoWorkItemProvider.ts [hunk 10]
```diff
@@ -118,60 +117,62 @@ export class AdoWorkItemProvider extends BaseProvider {
...
-    results.forEach((result, index) => {
-      if (result.status === 'fulfilled') {
-        const { items, failed } = result.value;
-        allItems.push(...items);
-        if (failed) {
-          failures.push(projectList[index] || this.org);
+      const projectList = validProjects.length > 0 ? validProjects : [''];
+      const results = await Promise.allSettled(
+        projectList.map(project => this.fetchWorkItemsForProject(accessToken, orgConfig.org, project)),
+      );
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoWorkItemProvider.ts:146]: Intentional design: sequential processing avoids overwhelming the ADO API with concurrent requests across orgs. Will consider parallelizing if latency becomes an issue.

File: packages/ado/src/adoPrReviewProvider.ts [hunk 11]
```diff
@@ -102,102 +101,115 @@ export class AdoPrReviewProvider extends BaseProvider {
...
-    results.forEach((result, index) => {
-      if (result.status === 'fulfilled') {
-        const { items, failed } = result.value;
-        allItems.push(...items);
-        if (failed) {
-          failures.push(projectList[index] || this.org);
+      const projectList = validProjects.length > 0 ? validProjects : [''];
+      const results = await Promise.allSettled(
+        projectList.map(project => this.fetchPrsForProject(accessToken, orgConfig.org, project, userId)),
+      );
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPrReviewProvider.ts:138]: Same as above — sequential processing is a deliberate choice to be conservative with API rate limits. Can optimize later if needed.

File: packages/ado/src/adoPrReviewProvider.ts [hunk 12]
```diff
@@ -102,102 +101,115 @@ export class AdoPrReviewProvider extends BaseProvider {
...
         {
           headers: { Authorization: `Bearer ${token}` },
         },
       );
     } catch (err) {
       logger.error('Network error fetching connection data:', err);
-      this._cachedUserId = undefined;
-      this._cachedSessionAccountId = undefined;
+      this._cachedUserIds.delete(org);
       return undefined;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPrReviewProvider.ts:206]: Fixed in 5265d7d

File: packages/ado/src/adoPrReviewProvider.ts [hunk 13]
```diff
@@ -102,102 +101,115 @@ export class AdoPrReviewProvider extends BaseProvider {
...
     }
 
     if (!response.ok) {
       const body = await response.text().catch(() => '');
       logger.error(`Failed to fetch connection data: ${response.status} ${body}`);
-      this._cachedUserId = undefined;
-      this._cachedSessionAccountId = undefined;
+      this._cachedUserIds.delete(org);
       return undefined;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPrReviewProvider.ts:213]: Fixed in 5265d7d

File: packages/ado/src/adoWorkItemProvider.ts [hunk 14]
```diff
@@ -267,13 +269,13 @@ export class AdoWorkItemProvider extends BaseProvider {
...
 
     const items: DiscoveredItem[] = activeWorkItems.map((wi) => {
       const projectName = wi.fields['System.TeamProject'];
       const wiType = wi.fields['System.WorkItemType'];
       return {
-        externalId: `${projectName}/${wi.id}`,
+        externalId: `${org}/${projectName}/${wi.id}`,
         title: `${wiType} ${wi.id}: ${wi.fields['System.Title']}`,
         description: wi.fields['System.Description']?.replace(/<[^>]*>/g, '')?.slice(0, 200),
         url: wi._links.html.href,
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoWorkItemProvider.ts:281]: Intentional: the group is a display label in the UI tree. Including org would make it verbose for the common single-org case. Cross-org collisions are unlikely in practice.

File: packages/ado/src/adoWorkItemProvider.ts [hunk 15]
```diff
@@ -200,13 +202,13 @@ export class AdoWorkItemProvider extends BaseProvider {
...
+      logger.error(`Network error querying work items for project "${project || org}":`, err);
       return { items: [], failed: true };
     }
 
     if (!wiqlResponse.ok) {
-      logger.warn(`Failed to fetch work items for project: ${project || this.org}`);
-      logger.error(`WIQL query failed for project "${project || this.org}": ${wiqlResponse.status}`);
+      logger.warn(`Failed to fetch work items for project: ${project || org}`);
+      logger.error(`WIQL query failed for project "${project || org}": ${wiqlResponse.status}`);
       return { items: [], failed: true };
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoWorkItemProvider.ts:212]: Fixed in 2285b0b

File: packages/ado/src/adoWorkItemProvider.ts [hunk 16]
```diff
@@ -230,7 +232,7 @@ export class AdoWorkItemProvider extends BaseProvider {
 
     for (let i = 0; i < ids.length; i += batchSize) {
       const batchIds = ids.slice(i, i + batchSize);
-      const detailUrl = `https://dev.azure.com/${encodeURIComponent(this.org)}/_apis/wit/workitems?ids=${batchIds.join(',')}&fields=System.Title,System.Description,System.TeamProject,System.WorkItemType,System.State&$expand=links&api-version=7.1`;
+      const detailUrl = `https://dev.azure.com/${encodeURIComponent(org)}/_apis/wit/workitems?ids=${batchIds.join(',')}&fields=System.Title,System.Description,System.TeamProject,System.WorkItemType,System.State&$expand=links&api-version=7.1`;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoWorkItemProvider.ts:236]: Fixed in 2285b0b

File: packages/ado/src/extension.ts [hunk 17]
```diff
@@ -69,28 +70,32 @@ export async function activate(_context: vscode.ExtensionContext): Promise<void>
...
 
-    if (!org) {
-      logger.info('No organization configured — set workcenterAdo.organization to enable ADO providers');
+    const orgConfigs = parseAdoProjectsConfig(projects, legacyOrg);
+
+    if (orgConfigs.length === 0) {
+      logger.info('No organizations configured — set workcenterAdo.projects to enable ADO providers');
       if (!orgWarningShown) {
-        vscode.window.showWarningMessage('WorkCenter ADO: Azure DevOps organization not configured. Set workcenterAdo.organization in settings.');
+        vscode.window.showWarningMessage(
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/extension.ts]: Fixed in 79ec29d

File: packages/ado/src/adoPrReviewProvider.ts [hunk 18]
```diff
@@ -102,135 +101,146 @@ export class AdoPrReviewProvider extends BaseProvider {
...
     }
 
-    this._cachedUserId = data.authenticatedUser.id;
-    this._cachedSessionAccountId = sessionAccountId;
-    logger.debug(`Resolved user ID: ${this._cachedUserId}`);
-    return this._cachedUserId;
+    this._cachedUserIds.set(org, data.authenticatedUser.id);
+    logger.debug(`Resolved user ID: ${data.authenticatedUser.id}`);
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPrReviewProvider.ts]: Fixed in 8130e13

File: packages/ado/src/adoWorkItemProvider.ts [hunk 19]
```diff
@@ -118,60 +117,62 @@ export class AdoWorkItemProvider extends BaseProvider {
...
-    for (const project of this.projects) {
-      if (project === '' || isValidUrlSegment(project)) {
-        validProjects.push(project);
-      } else {
-        logger.warn('Skipping invalid ADO project name', project);
+    for (const orgConfig of this.orgConfigs) {
+      if (!isValidUrlSegment(orgConfig.org)) {
+        logger.warn('Skipping fetch: invalid ADO organization name', orgConfig.org);
+        continue;
       }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoWorkItemProvider.ts:127]: Existing tests exercise the org iteration loop via the OrgConfig[] constructor. The configEdgeCases tests also cover multi-org parsing. Adding dedicated multi-org integration tests is a good follow-up but orthogonal to this PR's scope.

File: packages/ado/src/adoPrReviewProvider.ts [hunk 20]
```diff
@@ -102,135 +101,146 @@ export class AdoPrReviewProvider extends BaseProvider {
...
-    const userId = await this.getUserId(accessToken, sessionAccountId);
-    if (!userId) {
-      const message = 'Failed to determine Azure DevOps user identity';
-      if (isUserTriggered) {
-        void vscode.window.showWarningMessage(`WorkCenter ADO: ${message}`);
+    for (const orgConfig of this.orgConfigs) {
+      if (!isValidUrlSegment(orgConfig.org)) {
+        logger.warn('Skipping PR fetch: invalid ADO organization name', orgConfig.org);
+        continue;
       }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPrReviewProvider.ts:112]: Same as above - the core multi-org logic is tested through existing tests that use the OrgConfig[] constructor. Multi-org integration tests would be a good follow-up.

File: packages/ado/src/test/configParser.test.ts [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+      expect(result).toEqual<OrgConfig[]>([
+        { org: 'myorg', projects: [] },
+      ]);
+    });
+
+    it('org-only entry before specific projects still monitors whole org', () => {
+      const result = parseAdoProjectsConfig(['myorg', 'myorg/Proj1'], '');
+      expect(result).toEqual<OrgConfig[]>([
+        { org: 'myorg', projects: [] },
+      ]);
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/test/configParser.test.ts]: Fixed in 5acd7b3

File: packages/ado/src/configParser.ts [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+): OrgConfig[] {
+  const trimmedLegacyOrg = legacyOrganization.trim();
+
+  // Determine whether entries use new org/project format by checking for
+  // valid slash-separated entries (ignoring malformed ones like "/" or "/proj")
+  const hasNewFormatEntry = projects.some(p => {
+    const t = p.trim();
+    const idx = t.indexOf('/');
+    return idx > 0 && idx < t.length - 1;
+  });
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/configParser.ts]: Fixed in 09d5ca9

File: packages/ado/src/extension.ts [hunk 23]
```diff
@@ -69,28 +70,32 @@ export async function activate(_context: vscode.ExtensionContext): Promise<void>
...
-      logger.info('No organization configured — set workcenterAdo.organization to enable ADO providers');
+    const orgConfigs = parseAdoProjectsConfig(projects, legacyOrg);
+
+    if (orgConfigs.length === 0) {
+      logger.info('No organizations configured — set workcenterAdo.projects to enable ADO providers');
       if (!orgWarningShown) {
-        vscode.window.showWarningMessage('WorkCenter ADO: Azure DevOps organization not configured. Set workcenterAdo.organization in settings.');
+        void vscode.window.showWarningMessage(
+          'WorkCenter ADO: No Azure DevOps organizations configured. Add entries to workcenterAdo.projects (e.g. "myorg" or "myorg/myproject").',
+        );
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/extension.ts]: Fixed in 26ec38b

File: packages/ado/package.json [hunk 24]
```diff
@@ -22,18 +22,13 @@
...
         "workcenterAdo.projects": {
           "type": "array",
           "items": {
             "type": "string"
           },
           "default": [],
-          "description": "Azure DevOps projects to monitor. Leave empty to fetch assigned work items and PRs across the entire organization."
+          "description": "Azure DevOps organizations and projects to monitor. Each entry is either '<org>' (monitor entire organization) or '<org>/<project>' (monitor a specific project). Supports multiple organizations."
         },
         "workcenterAdo.refreshIntervalSeconds": {
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/package.json:33]: Intentional — backward compatibility for the deprecated organization setting has been removed per updated requirements. The PR description has been updated to reflect this breaking change.

File: packages/ado/src/extension.ts [hunk 25]
```diff
@@ -69,28 +70,42 @@ export async function activate(_context: vscode.ExtensionContext): Promise<void>
...
     const projects = config.get<string[]>('projects', []);
 
-    if (!org) {
-      logger.info('No organization configured — set workcenterAdo.organization to enable ADO providers');
-      if (!orgWarningShown) {
-        vscode.window.showWarningMessage('WorkCenter ADO: Azure DevOps organization not configured. Set workcenterAdo.organization in settings.');
-        orgWarningShown = true;
+    const orgConfigs = parseAdoProjectsConfig(projects);
+
+    if (orgConfigs.length === 0) {
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/extension.ts:77]: Intentional — backward compatibility for the deprecated organization setting has been removed per updated requirements. The PR description has been updated to reflect this breaking change.

File: packages/ado/src/configParser.ts [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * Malformed entries such as `/`, `org/`, `/project`, and multi-slash entries
+ * like `org/proj/extra` are silently skipped.
+ */
+export function parseAdoProjectsConfig(
+  projects: string[],
+): OrgConfig[] {
+  if (projects.length === 0) {
+    return [];
+  }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/configParser.ts:26]: Intentional — backward compatibility for the deprecated organization setting has been removed per updated requirements. The PR description has been updated to reflect this breaking change.
